CMakeLists.txt

```diff
@@ -1,61 +1,179 @@
+option(QL_USE_INDEXED_COUPON "Use indexed coupons instead of par coupons" OFF)
+option(QL_USE_STD_CLASSES "Enable all QL_USE_STD_ options" OFF)
+option(QL_USE_STD_SHARED_PTR "Use standard smart pointers instead of Boost ones" OFF)
+option(QL_USE_STD_UNIQUE_PTR "Use std::unique_ptr instead of std::auto_ptr" ON)
+option(QL_USE_STD_FUNCTION "Use std::function and std::bind instead of Boost ones" OFF)
+option(QL_USE_STD_TUPLE "Use std::tuple instead of boost::tuple" OFF)
+
+# Convenience option to activate all STD options
+if (QL_USE_STD_CLASSES)
+    set(QL_USE_STD_SHARED_PTR ON)
+    set(QL_USE_STD_UNIQUE_PTR ON)
+    set(QL_USE_STD_FUNCTION ON)
+    set(QL_USE_STD_TUPLE ON)
+endif()
+
+# Sessions requires thread-safe initialization
+if (QL_ENABLE_SESSIONS)
+    set(QL_ENABLE_SINGLETON_THREAD_SAFE_INIT ON)
+endif()
+
+# Shared or static libs
+if (NOT DEFINED BUILD_SHARED_LIBS)
+    set(BUILD_SHARED_LIBS OFF)
+endif()
+
+# Boost thread and unit_test_framework needed for tests and benchmark
+find_package(Boost 1.58.0 REQUIRED COMPONENTS thread unit_test_framework)
+
+# Avoid using Boost auto-linking
+add_compile_definitions(BOOST_ALL_NO_LIB)
 
-    # Remove warnings
-    add_definitions(-D_SCL_SECURE_NO_WARNINGS -D_CRT_SECURE_NO_WARNINGS)
+if (QL_ENABLE_OPENMP)
+    find_package(OpenMP REQUIRED)
 endif()
 
-# On single-configuration builds, select a default build type that
-# gives the same compilation flags as a default autotools build.
-if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
-    set(CMAKE_BUILD_TYPE "RelWithDebInfo")
+# Parallel test runner needs library rt on *nix for shm_open, etc.
+if (QL_ENABLE_PARALLEL_UNIT_TEST_RUNNER AND UNIX AND NOT APPLE)
+    find_library(RT_LIBRARY rt REQUIRED)
 endif()
 
-# to reference headers via <ql/foo.hpp>, we need to add the root
-# directory of the project to includes
-include_directories(${CMAKE_CURRENT_SOURCE_DIR})
+# Require C++11 or higher
+if (NOT DEFINED CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_STANDARD 11)
+elseif(CMAKE_CXX_STANDARD LESS 11)
+    message(FATAL_ERROR "Please specify CMAKE_CXX_STANDARD of 11 or higher")
+endif()
+if (NOT DEFINED CMAKE_CXX_STANDARD_REQUIRED)
+    set(CMAKE_CXX_STANDARD_REQUIRED ON)
+endif()
+# Avoid use of compiler language extensions, i.e. -std=c++11 not -std=gnu++11
+if (NOT DEFINED CMAKE_CXX_EXTENSIONS)
+    set(CMAKE_CXX_EXTENSIONS FALSE)
+endif()
+
+# If available, use PIC for shared libs and PIE for executables
+if (NOT DEFINED CMAKE_POSITION_INDEPENDENT_CODE)
+    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endif()
+if (CMAKE_POSITION_INDEPENDENT_CODE)
+    # cmake policy CMP0083: add PIE support if possible (need cmake 3.14)
+    include(CheckPIESupported)
+    check_pie_supported()
+endif()
+
+# Configure files
+set(QL_VERSION ${PACKAGE_VERSION})
+set(QL_HEX_VERSION ${PACKAGE_VERSION_HEX})
+configure_file(ql/config.hpp.cfg ql/config.hpp @ONLY)
+configure_file(ql/version.hpp.cfg ql/version.hpp @ONLY)
+add_compile_definitions(HAVE_CONFIG_H)
```

Comment:
There's a problem with this define that we also had in the autotools build — here it's defined automatically, but once the compilation is done and the library is installed, client code will need to define it in order to pick up `config.hpp` and the option definitions in there.  In the autotools run, when we install, we execute the code at https://github.com/lballabio/QuantLib/blob/master/ql/Makefile.am#L163 to (a) rename it because it's a common define name and (b) define it in the installed files themselves.  I'm not sure what's possible here?

Reply:
You're right.  I missed that because most libraries built with autotools use `HAVE_CONFIG_H` internally only, i.e.

```cpp
// some_library_unit.cpp
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
```

Placement in the client headers propagates that upward to the clients.   I'll commit a reasonable  fix.